<launch>

    <arg name="robot_ip"                                default="172.16.0.2" />
    <arg name="table_height"                            default="" doc="Unset, if present will override the grasp_server_config YAML"/>
    <arg name="camera_width"                            default="1280" />
    <arg name="camera_height"                           default="720" />
    <arg name="camera_fps"                              default="15" />
    <arg name="rviz_config"                             default="calib_reach.rviz" />
    <arg name="grasp_server_config"                     default="grasp_server_config.yaml" />
    <arg name="hand_camera_serial"                      doc="Serial number of the hand-mounted camera" />
    <arg name="setup_camera_serial"                     doc="Serial number of the setup-mounted camera" />


    <!-- Start hand-mounted realsense -->
    <include file="$(find realsense2_camera)/launch/rs_camera.launch" >
        <arg name="camera"                              value="hand_camera" />
        <arg name="tf_prefix"                           value="hand_camera" />
        <arg name="serial_no"                           value="$(arg hand_camera_serial)" />
        <arg name="enable_depth"                        value="false" />
        <arg name="color_width"                         value="$(arg camera_width)" />
        <arg name="color_height"                        value="$(arg camera_height)" />
        <arg name="color_fps"                           value="$(arg camera_fps)" />
    </include>

    <!-- Temporary: fix the hand_camera_link to camera_link -->
    <!-- TODO: fix namespaces and get rid of this -->
    <node pkg="tf" type="static_transform_publisher" name="hand_cam_fix" args="0 0 0 0 0 0 camera_link hand_camera_link 100"/>

    <!-- Start setup-mounted realsense -->
    <include file="$(find realsense2_camera)/launch/rs_camera.launch" >
        <arg name="camera"                              value="setup_camera" />
        <arg name="tf_prefix"                           value="setup_camera" />
        <arg name="serial_no"                           value="$(arg setup_camera_serial)" />
        <arg name="enable_depth"                        value="false" />
        <arg name="color_width"                         value="$(arg camera_width)" />
        <arg name="color_height"                        value="$(arg camera_height)" />
        <arg name="color_fps"                           value="$(arg camera_fps)" />
    </include>

    <!-- Start multiplexer nodes -->
    <node name="camera_info_mux" pkg="topic_tools" type="mux" args="camera/color/camera_info hand_camera/color/camera_info setup_camera/color/camera_info mux:=camera_info_mux" />
    <node name="camera_image_mux" pkg="topic_tools" type="mux" args="camera/color/image_raw hand_camera/color/image_raw setup_camera/color/image_raw mux:=camera_image_mux" />

    <!-- Switch camera with
    rosservice call /camera_info_mux/select "topic: '/setup_camera/color/camera_info'"
    rosservice call /camera_image_mux/select "topic: '/setup_camera/color/image_raw'"
    -->

    <!-- Start the aruco board detection node -->
    <include file="$(find aruco_board_detect)/launch/aruco_board_detect.launch" >
        <arg name="camera_info_topic"                   value="/camera/color/camera_info" />
        <arg name="camera_image_topic"                  value="/camera/color/image_raw" />
        <arg name="publish_single_markers"              value="true" />
        <arg name="detection_rate"                      value="0.5" />
    </include>

    <!-- Start robot controllers, moveit and rviz -->
    <include file="$(find panda_moveit_config)/launch/panda_control_moveit_rviz.launch" >
        <arg name="load_gripper"                    value="true" />
        <arg name="robot_ip"                        value="$(arg robot_ip)" />
        <arg name="hand_camera"                     value="true" />
        <arg name="rviz_command_args"               value="-d $(find panda_grasp_server)/rviz/$(arg rviz_config)" />
    </include>

    <!-- Start the grasp and movement server node -->
    <rosparam command="load" file="$(find panda_grasp_server)/config/$(arg grasp_server_config)" ns="panda_grasp_server" />
    <param name="/panda_grasp_server/workspace/table_height" type="double" value="$(arg table_height)" unless="$(eval table_height == '')" />
    <node name="panda_grasp_server" pkg="panda_grasp_server" type="panda_grasp_server_node" output="log" />

    <!-- Start the calibration_reachability script -->
    <node name="calibration_reachability" pkg="panda_grasp_server" type="graspa_calibration_reachability.py" output="screen" />

</launch>
